server/src-lib/Hasura/GraphQL/Resolve/Insert.hs

```diff
@@ -193,6 +196,12 @@ parseOnConflict tn updFiltrM allColMap val = withPathK "on_conflict" $
       (_, enumVal) <- asEnumVal v
       return $ ConstraintName $ G.unName $ G.unEnumValue enumVal
 
+    parseWhereExp o = do
+      whereExpM <- forM (OMap.lookup "where" o) $ \v -> do
+        unresolved <- parseBoolExp v
+        traverse (traverse resolveValTxt) unresolved
```

Comment:
I’d probably write this as `traverse (traverse resolveValTxt) =<< parseBoolExp v` so you don’t have to come up with an intermediate name (since `unresolved` isn’t very meaningful, anyway). In fact, you could even get rid of the lambda entirely and write

```haskell
for (OMap.lookup "where" o) $
  traverse (traverse resolveValTxt) <=< parseBoolExp
```

to eliminate the equally-unhelpful `v` name… and if you wanted to take it *really* far, you could rewrite the whole thing point-free in a sort of “pipeline” style, like

```haskell
parseWhereExp =
      OMap.lookup "where"
  >>> traverse (parseBoolExp >=> traverse (traverse resolveValTxt))
  >>> fmap (maybe (S.BELit True) (toSQLBoolExp (S.mkQual tn)))
```

but I’m not sure that’s actually better. :) I’ll leave it up to you if you want to incorporate any of these changes, or how far you want to go with them.